### ISSUE TITLE
Illuminated Cloud-like Live code templates

### ISSUE THREAD

**Issue by osieckiAdam** [forcedotcom/salesforcedx-vscode#1555]
**Describe the solution you'd like**
It would be very cool to have live code templates like in the Illuminated Cloud plugin for writing APEX code. Currently we are able to create snippets in Emmet, which is very helpful, but we don't have access to SObject types, SObject fields, so we are not able to create snippets like SEL*, dfr which are metadata-aware

**Additional context**
IC Live templates are described here:
http://www.illuminatedcloud.com/home/completion/livetemplates

**Comment by ChuckJonas** [comment]
when you say they are "metadata aware", what exactly do you mean?

https://github.com/ChuckJonas/vscode-apex-snippets

I've started a git repository for common apex snippets.  I planned to submit a PR once I've had a chance to clean them up a bit.

**Comment by osieckiAdam** [comment]
In Illuminated Cloud You are able to generate 'Offline Symbol Table' which basically is a similiar functionality to SFDX: Refresh SObject Definitions. It allows to use names of standard and custom object and field names in apex code completion. But in IC it also can be used in snippets. 

For example: You can type SEL* snippet, and it will allow You to choose between all standards and custom objects in Your org. For the chosen object it will put all standard and custom fields in SELECT part of the query. You can check the details of live templates in IC in my first comment

**Comment by jefersonchaves** [comment]
Hey @osieckiAdam: IC Templates looks a lot like [VS Code Snippets](https://code.visualstudio.com/docs/editor/userdefinedsnippets). Which I guess is what @ChuckJonas is referring to. Is that similar?

**Comment by osieckiAdam** [comment]
Yes it is similar, but for now I couldn't find the way how to create snippets which are using variables like allFields (see SEL* snippet in IC), SObjectType (dfr in IC), fields (sqit in IC)

**Comment by ChuckJonas** [comment]
Unfortunately the vscode snippet engine doesn't handle use cases like this at the moment.  I don't see anything in the extension API that would allow you to provide these capabilities  (short of writing a completely custom snippet engine).
